hubble: add SNAT IP flow field and filter

kaworu opened the pull request:
<del>_NOTE: draft until https://github.com/cilium/cilium/pull/31226 is merged, also currently missing the Hubble CLI bits_</del>

Add a new `source_xlated` field to Hubble `Flow.IP` and a corresponding IP filter.

Fix the `FlowFilter.Experimental.cel_expression` and `FlowFilter.drop_reason_desc` protobuf field numbers on the way.

kaworu committed:
hubble: fix traffic direction for lb and egress gw

In an egress gateway setup, the packet leaving the egress gateway to the
destination is SNAT'ed with the egress gateway IP address.

Before this patch, the corresponding flow would be reported wrongly as
INGRESS by Hubble, as the traffic direction heuristic would see a
non-reply flow emitted by a different BPF program that the one attached
to the source lxc interface.

This patch set the traffic direction to EGRESS when we detect that the
packet was SNAT'ed, which also fix the lb case.

Signed-off-by: Alexandre Perrin <alex@isovalent.com>

kaworu committed:
hubble: refactor threefour parser Decode

Move the ip != nil check earlier since the block is about setting ip
related fields and the check is repeated before this patch. This will
simplify the next commit introducing new logic related to ip.Source.

Signed-off-by: Alexandre Perrin <alex@isovalent.com>

kaworu committed:
hubble: add IP.source_xlated

Before this patch, when TraceNotify.OriginalIP was set, the Hubble
parser would use it as ip.Source discarding the source IP parsed from
the packet header.

This patch introduce a new IP.source_xlated field, preserving the parsed
header's source IP, allowing to expose the actual IP used in the Egress
Gateway and LB cases.

Signed-off-by: Alexandre Perrin <alex@isovalent.com>

kaworu committed:
hubble: fix FlowFilter protobuf field numbers

Since cel_expression is part of the new Experimental message, it doesn't
have to follow the FlowFilter field number.

Also fix drop_reason_desc field to use the highest unused field number.

Both fields were introduced recently (i.e. after v1.15) and thus this
commit doesn't break backward compatibility with any currently released
Cilium version at the time of writing.

Signed-off-by: Alexandre Perrin <alex@isovalent.com>

kaworu committed:
hubble: add source_ip_xlated filter

The previous commit introduce a new SNAT IP field (source_xlated) in
Hubble flow data. This commit is a follow-up adding a corresponding
Hubble filter by SNAT IP either with exact match or CIDR match.

Note that unlike source and destination IP, source_xlated might be
empty.

Signed-off-by: Alexandre Perrin <alex@isovalent.com>

chancez commented on api/v1/flow/flow.proto:
```diff
@@ -258,6 +272,7 @@ message TCP {
 
 message IP {
     string source = 1;
+    string snatAs = 5;
```
A few nits:
- Should this be snake case? Would be easier to read.
- The name is a bit unobvious to me. Maybe `source_nat` or `original_source` or something? (I'm not sure if this the IP pre or post translation, but if it's post translation it could be `new_source`).
- Should probably add a comment to describe the field's purpose.

kaworu replied:
Hey @chancez thanks for chiming in, no worries this PR is in draft because it's on top of https://github.com/cilium/cilium/pull/31226 but I opened it anyway exactly to get that kind of feedback :pray:

* snake case: yes, actually that is what [the protobuf style guide recommends](https://protobuf.dev/programming-guides/style/#message-field-names) but for some reason I got mixed up and thought camel case was to be preferred, so I'll update.
* The name is not ideal I agree, I couldn't come up with something better so I'm open to suggestions. To clarify: in the SNAT case `IP.source` is the pre translation IP and `snatAs` the post-translation IP. With that in mind, `source_nat` is ambiguous and `original_source` confusing IMHO, but let's hear more opinions on this.
* comment: yes.

chancez replied:
Hmmm yeah. Is `post_translate_blahblah` too much? Naming things is hard.

kaworu replied:
`source_xlated` or `source_xlated_as` ? Let's figure out the name before I complete the Hubble CLI part.

chancez replied:
`source_xlated` or `source_translated` seems decent

kaworu replied:
Went with `source_xlated`, please take another look :pray:

chancez reviewed (changes requested):
Could use a parser test where `SnatAs` is non-empty.

chancez commented on api/v1/flow/flow.proto:
```diff
@@ -521,7 +542,7 @@ message FlowFilter {
     repeated Verdict verdict = 5;
     // only applicable to Verdict = DROPPED (e.g. "POLICY_DENIED", "UNSUPPORTED_L3_PROTOCOL")
-    repeated DropReason drop_reason_desc = 35;
+    repeated DropReason drop_reason_desc = 33;
```
Isn't this a breaking change?

kaworu replied:
Strictly speaking yes, it is a breaking change. In practice there are no Cilium release yet with this field, so it should be safe to go. See [this tread](https://github.com/cilium/cilium/pull/32135#discussion_r1580739954) on the PR introducing the `drop_reason_desc` field and the commit message:
> Both fields were introduced recently (i.e. after v1.15) and thus this
commit doesn't break backward compatibility with any currently released
Cilium version at the time of writing.

chancez replied:
Oh right; I forgot this was new in the filters.

chancez commented on pkg/hubble/parser/threefour/parser_test.go:
```diff
@@ -160,6 +161,7 @@ func TestL34Decode(t *testing.T) {
 
 	assert.Equal(t, []string{"host-192.168.60.11"}, f.GetSourceNames())
 	assert.Equal(t, "192.168.60.11", f.GetIP().GetSource())
+	assert.Equal(t, flowpb.TraceReason_ESTABLISHED, f.GetTraceReason())
```
I think we still need a test that asserts the value of `f.GetIP.GetSourceXlated()` after parsing.

kaworu replied:
Sure that can't hurt. For context `TestL34Decode` is mostly useful to ensure we're not breaking the `TraceNotify` field layout by accident. It should be parsed as a `TraceNotifyV0` struct which doesn't have `OrigIP` and thus can't have `SourceXlated` set, and this part is already tested in `TestTraceNotifyOriginalIP`.

kaworu replied:
Added `f.GetIP().GetSourceXlated()` checks in `TestL34Decode` :pray:

kaworu committed:
hubble/cli: add --snat-ip

The previous commits introduced the IP.source_xlated flow field and a
corresponding source_ip_xlated filter at the Hubble API level. This
patch is a follow-up adding source_ip_xlated support to the Hubble CLI.

After this patch, the --ip flag behaviour is arguably unintuitive as it
will not take source_ip_xlated into account (i.e. only source_ip and
destination_ip).  Unfortunately, addressing this limitation would
require reworking the Hubble CLI internal left/right filter setup, which
it outside of the scope of this commit. The --ip help message was
adapted accordingly with the intent of clearing the potential confusion.

Signed-off-by: Alexandre Perrin <alex@isovalent.com>

chancez reviewed (approved):
LGTM

gentoo-root commented on hubble/cmd/observe/flows_filter.go:
```diff
@@ -189,6 +189,7 @@ func newFlowFilter() *flowFilter {
 			{"from-pod", "namespace", "all-namespaces"},
 			{"to-service", "namespace", "all-namespaces"},
 			{"from-service", "namespace", "all-namespaces"},
+			{"snat-ip"},
```
Maybe I'm missing something, but what's the point of conflict lists that consist of one item? What do they check?

rolinh replied:
That's a good question :sweat_smile: I think there's no point in doing that although we apparently did it for other filters as well.

gentoo-root reviewed (approved):
Looks good, one question.
